datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java

```diff
@@ -0,0 +1,132 @@
+package com.sequenceiq.datalake.service.sdx.attach;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+import com.sequenceiq.authorization.service.OwnerAssignmentService;
+import com.sequenceiq.cloudbreak.common.service.TransactionService;
+import com.sequenceiq.datalake.entity.DatalakeStatusEnum;
+import com.sequenceiq.datalake.entity.SdxCluster;
+import com.sequenceiq.datalake.repository.SdxClusterRepository;
+import com.sequenceiq.datalake.service.sdx.status.SdxStatusService;
+
+/**
+ * Provides core logic and functionality for attaching or reattaching an SDX cluster.
+ */
+@Component
+public class SdxAttachService {
+    private static final Logger LOGGER = LoggerFactory.getLogger(SdxAttachService.class);
+
+    @Inject
+    private TransactionService transactionService;
+
+    @Inject
+    private SdxClusterRepository sdxClusterRepository;
+
+    @Inject
+    private OwnerAssignmentService ownerAssignmentService;
+
+    @Inject
+    private SdxDetachNameGenerator sdxDetachNameGenerator;
+
+    @Inject
+    private SdxStatusService sdxStatusService;
+
+    @Inject
+    private SdxAttachDetachUtils sdxAttachDetachUtils;
+
+    /**
+     * Can throw the following exceptions:
+     *      - Exceptions thrown by reattachCluster.
+     *      - Exceptions thrown by reattachStack.
+     *      - Exceptions thrown by reattachExternalDatabase.
+     *      - Exceptions thrown by attachSdx.
+     */
+    public SdxCluster reattachDetachedSdxCluster(SdxCluster clusterToReattach) throws Exception {
+        String detachedName = clusterToReattach.getClusterName();
+        String detachedCrn = clusterToReattach.getCrn();
+
+        return transactionService.required(() -> {
+            SdxCluster reattached = reattachCluster(clusterToReattach);
+            reattachStack(reattached, detachedName);
+
+            if (clusterToReattach.hasExternalDatabase()) {
+                reattachExternalDatabase(reattached, detachedCrn);
```

Comment:
external calls shouldn't be part of a transaction as it could cause long db locks

Reply:
Fixed.